tracee/pipeline.go

```diff
@@ -0,0 +1,231 @@
+			if !t.shouldPrintEvent(rawEvent.Ctx.EventID) {
+				continue
+			}
+			err := t.prepareArgsForPrint(&rawEvent.Ctx, rawEvent.RawArgs)
+			if err != nil {
+				errc <- err
+				continue
+			}
+			args := make([]interface{}, rawEvent.Ctx.Argnum)
+			argsNames := make([]string, rawEvent.Ctx.Argnum)
+			for i, tag := range rawEvent.ArgsTags {
+				args[i] = rawEvent.RawArgs[tag]
+				argName, ok := argNames[tag]
+				if ok {
+					argsNames[i] = argName
+				} else {
+					errc <- fmt.Errorf("Invalid arg tag")
+					continue
+				}
+			}
+			t.stats.eventCounter.Increment()
+			evt, err := newEvent(rawEvent.Ctx, argsNames, args)
+			if err != nil {
+				errc <- err
+				continue
+			}
+			select {
+			case out <- evt:
+			case <-done:
+				return
+			}
+		}
+	}()
+	return out, errc, nil
+}
+
+func (t *Tracee) printEvent(done <-chan struct{}, in <-chan Event) (<-chan error, error) {
+	errc := make(chan error, 1)
+	go func() {
+		defer close(errc)
+		for printEvent := range in {
+			t.printer.Print(printEvent)
+		}
+	}()
+	return errc, nil
+}
+
+// WaitForPipeline waits for results from all error channels.
+func (t *Tracee) WaitForPipeline(errs ...<-chan error) error {
+	errc := MergeErrors(errs...)
+	for err := range errc {
+		t.handleError(err)
+	}
+	return nil
+}
+
+// MergeErrors merges multiple channels of errors.
+// Based on https://blog.golang.org/pipelines.
+func MergeErrors(cs ...<-chan error) <-chan error {
+	var wg sync.WaitGroup
+	// We must ensure that the output channel has the capacity to hold as many errors
+	// as there are error channels. This will ensure that it never blocks, even
+	// if WaitForPipeline returns early.
+	out := make(chan error, len(cs))
+
+	// Start an output goroutine for each input channel in cs.  output
+	// copies values from c to out until c is closed, then calls wg.Done.
+	output := func(c <-chan error) {
+		for n := range c {
+			out <- n
+		}
+		wg.Done()
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		go output(c)
+	}
+
+	// Start a goroutine to close out once all the output goroutines are
+	// done.  This must start after the wg.Add call.
+	go func() {
+		wg.Wait()
+		close(out)
+	}()
+	return out
+}
+
+/*
+// An example for splitting the pipeline
+// Based on https://medium.com/statuscode/pipeline-patterns-in-go-a37bb3a7e61d
+func splitter(ctx context.Context, in <-chan int64) (
+	<-chan int64, <-chan int64, <-chan error, error) {
+	out1 := make(chan int64)
+	out2 := make(chan int64)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(out1)
+		defer close(out2)
+		defer close(errc)
+		for n := range in {
+			// Send the data to the output channel 1 but return early
+			// if the context has been cancelled.
+			select {
+			case out1 <- n:
+			case <-done:
+				return
+			}
+			// Send the data to the output channel 2 but return early
+			// if the context has been cancelled.
+			select {
+			case out2 <- n:
+			case <-done:
+				return
+			}
+		}
+	}()
+	return out1, out2, errc, nil
+}
+*/
```

Comment:
I don't think this belongs here. maybe add as a comment to the PR if you want to save it for yourself in the context of the code